### PR TITLE
feat(ui): Add/open sudo modal when API endpoints require sudo

### PR DESCRIPTION
**Pull request by billyvg**
Requires https://github.com/getsentry/sentry/pull/6685

sudosm

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/api/endpoints/authenticator_index.py
- src/sentry/api/endpoints/sudo.py


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/static/sentry/app/__mocks__/api.jsx [hunk 1]
```diff
@@ -72,7 +88,14 @@ class Client {
...
       };
-      respond(Client.mockAsync, options.error, resp);
+      this.handleRequestError(
+        {
+          path: url,
+          requestOptions: options,
+        },
+        resp
+      );
+      // respond(Client.mockAsync, options.error, resp);
```
Comment (macqueen) [inline src/sentry/static/sentry/app/__mocks__/api.jsx]: forgotten comment

**Review by macqueen — COMMENTED** [review]
left a few comments. generally seems ok, but i'd like @mattrobenolt or someone else more familiar with auth stuff to review

File: src/sentry/api/endpoints/sudo.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from django.contrib import auth
+from rest_framework.response import Response
+from sudo.utils import grant_sudo_privileges
+
+from sentry.api.base import Endpoint
+from sentry.models import Authenticator
+
+
+class SudoEndpoint(Endpoint):
+    permission_classes = ()
```
[the remark this answers is not in the document]
Comment (mattrobenolt) [reply src/sentry/api/endpoints/sudo.py]: Yeah, it'll raise an `AttributeError`. I don't see any reason this wouldn't be restricted to `IsAuthenticated`.

File: src/sentry/api/endpoints/sudo.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        if 'challenge' in request.DATA and 'response' in request.DATA:
+            try:
+                interface = Authenticator.objects.get_interface(request.user, 'u2f')
+                if not interface.is_enrolled:
+                    raise LookupError()
+
+                challenge = json.loads(request.DATA['challenge'])
+                response = json.loads(request.DATA['response'])
+                authenticated = interface.validate_response(request, challenge, response)
+            except ValueError:
```
Comment (mattrobenolt) [inline src/sentry/api/endpoints/sudo.py]: You can catch both of these at the same time with `except (ValueError, LookupError)`

The only real comment I have here though is that this is potentially catching a wider group of errors than you'd want or expect. I'm not sure within this what you expect to trigger a `ValueError` either. And the concern here is that `LookupError` covers a lot of stuff. Including both `KeyError` and `AttributeError`.

Comment (billyvg) [reply src/sentry/api/endpoints/sudo.py]: I've updated to catch `LookupError` if u2f interface does not exist or user is not enrolled, and to catch `LookupError` only when decoding json.

**Comment by billyvg** [comment]
Addressed review issues + added some tests, can you re-review @mattrobenolt? Thanks

File: src/sentry/api/decorators.py [hunk 4]
```diff
@@ -27,7 +29,7 @@ def wrapped(self, request, *args, **kwargs):
                 "sudoRequired": True,
                 "username": request.user.username,
             }
-            return HttpResponse(json.dumps(data), status=401)
+            return Response(data, content_type="application/json", status=401)
```
Comment (mattrobenolt) [inline src/sentry/api/decorators.py]: nit: the content_type here isn't needed. It's default for JSON.

File: src/sentry/api/endpoints/sudo.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+from __future__ import absolute_import
+
+import json
```
Comment (mattrobenolt) [inline src/sentry/api/endpoints/sudo.py]: Use `sentry.utils.json` instead.

File: src/sentry/api/endpoints/sudo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattrobenolt) [reply src/sentry/api/endpoints/sudo.py]: @billyvg It seems this still applies from what I can tell?

**Review by mattrobenolt — CHANGES_REQUESTED** [review]

File: src/sentry/api/endpoints/sudo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (billyvg) [reply src/sentry/api/endpoints/sudo.py]: Sorry, not sure why I missed this

File: src/sentry/api/endpoints/sudo.py [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
 from sudo.utils import grant_sudo_privileges
 
 from sentry.api.base import Endpoint
 from sentry.models import Authenticator
+from sentry.utils import json
 
 
 class SudoEndpoint(Endpoint):
-    permission_classes = ()
+    permission_classes = (IsAuthenticated)
```
Comment (mattrobenolt) [inline src/sentry/api/endpoints/sudo.py]: Need a comma. Otherwise this isn't a tuple.

**Review by mattrobenolt — APPROVED** [review]
Assuming tests pass... 🥇🔒

**Pull request merged**